exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/NestedLoopJoinTemplate.java

```diff
@@ -40,132 +41,133 @@
+        break;
+      }
+      // reset state and get next left batch
+      resetAndGetNextLeft();
+    }
+    return outputIndex;
+  }
+
+  /**
+   * This method is the core of the nested loop join.For each left batch record looks for matching record
+   * from the list of right batches. Match is checked by calling {@link #doEval(int, int, int)} method.
+   * If matching record is found both left and right records are written into output batch,
+   * otherwise if join type is LEFT, than only left record is written, right batch record values will be null.
+   *
+   * @param joinType join type (INNER or LEFT)
    * @param outputIndex index to start emitting records at
    * @return final outputIndex after producing records in the output batch
    */
-  private int populateOutgoingBatch(int outputIndex) {
-
-    // Total number of batches on the right side
-    int totalRightBatches = rightCounts.size();
-
-    // Total number of records on the left
-    int localLeftRecordCount = leftRecordCount;
-
-    /*
-     * The below logic is the core of the NLJ. To have better performance we copy the instance members into local
-     * method variables, once we are done with the loop we need to update the instance variables to reflect the new
-     * state. To avoid code duplication of resetting the instance members at every exit point in the loop we are using
-     * 'goto'
-     */
-    int localNextRightBatchToProcess = nextRightBatchToProcess;
-    int localNextRightRecordToProcess = nextRightRecordToProcess;
-    int localNextLeftRecordToProcess = nextLeftRecordToProcess;
-
-    outer: {
-
-      for (; localNextRightBatchToProcess< totalRightBatches; localNextRightBatchToProcess++) { // for every batch on the right
-        int compositeIndexPart = localNextRightBatchToProcess << 16;
-        int rightRecordCount = rightCounts.get(localNextRightBatchToProcess);
-
-        for (; localNextRightRecordToProcess < rightRecordCount; localNextRightRecordToProcess++) { // for every record in this right batch
-          for (; localNextLeftRecordToProcess < localLeftRecordCount; localNextLeftRecordToProcess++) { // for every record in the left batch
-
+  private int populateOutgoingBatch(JoinRelType joinType, int outputIndex) {
+    // copy index and match counters as local variables to speed up processing
+    int nextRightBatchToProcess = tracker.getNextRightBatchToProcess();
+    int nextRightRecordToProcess = tracker.getNextRightRecordToProcess();
+    int nextLeftRecordToProcess = tracker.getNextLeftRecordToProcess();
+    boolean rightRecordMatched = tracker.isRightRecordMatched();
+
+    outer:
+    // for every record in the left batch
+    for (; nextLeftRecordToProcess < leftRecordCount; nextLeftRecordToProcess++) {
+      // for every batch on the right
+      for (; nextRightBatchToProcess < rightCounts.size(); nextRightBatchToProcess++) {
+        int rightRecordCount = rightCounts.get(nextRightBatchToProcess);
```

Comment:
Does that mean the query profile correctly show NxM rows processed in the nested loop join ?